### ISSUE TITLE
resolver should return directly, when use a `ip address` as a argument?

### ISSUE THREAD

**Issue by cssivision** [bluejekyll/trust-dns#197]
i do not see similar logic in the code, this can avoid useless dns query.

**Comment by bluejekyll** [comment]
I’m not sure if that is best practice or not. I understand the desire. Do you know how other resolvers deal with that case?

**Comment by cssivision** [comment]
extreme pursuit for performance...

**Comment by cssivision** [comment]
if we decide to do this, i can fire a pr.

**Comment by bluejekyll** [comment]
So the reason I'm concerned about this is that it's not clear to me that it's the responsibility of the resolver. i.e. I feel like, if you ask for name resolution, you expect it to be resolved :)

One option is that we could restrict the interface to `TryInto<Name>`, which I've been thinking of doing. That might make this more clear? For example, reverse Ip lookups are restricted to `IpAddr`.

I'm open to it, I just want to know what other resolvers out there do in this case...

**Comment by cssivision** [comment]
Now, if i use `ip addr` as input will get `Err` as a return value. 

i investigate the golang std lib method `net.LookupIP` i found it actually not implement the query, it just return the `ip` directly.

**Comment by cssivision** [comment]
Also it's hard for me to organize the code, for example:

I want to write code like this: 
```rust
// variable `host` may be a `domain` or `ip addr`.
let look_up = resolver.lookup_ip(&host);
let pair = look_up.and_then(move |res| {
      let addr = res.iter().next().expect("no addresses returned!");
      println!("{}", addr);
      TcpStream::connect(&SocketAddr::new(addr, port), &handle1).map(|c2| (c1, c2))
});
```

But unfortunately, i have to write code like this (still compile fail, even write like this): 
```rust
if let Ok(addr) = IpAddr::from_str(&host) {
      TcpStream::connect(&SocketAddr::new(addr, port), &handle1).map(|c2| (c1, c2))
} else {
      let look_up = resolver.lookup_ip(&host);
      let pair = look_up.and_then(move |res| {
            let addr = res.iter().next().expect("no addresses returned!");
            println!("{}", addr);
            TcpStream::connect(&SocketAddr::new(addr, port), &handle1).map(|c2| (c1, c2))
       });
       pair
};
```

it seems that i can't write code in async using future, when `host` is a condition to decide whether to do dns query.

**Comment by bluejekyll** [comment]
Ok. This makes sense. If you're still interested in putting that PR together, that would be awesome!

**Comment by cssivision** [comment]
ok, i will spend some time on this.

**Comment by cssivision** [comment]
sorry for being so late

**Comment by bluejekyll** [comment]
No worries! Thanks for putting it together!

**Comment by bluejekyll** [comment]
Fixed in #302
